Tensile/SolutionStructs.py

```diff
@@ -44,12 +46,277 @@ def pvar(state, field):
+    if self.packSpatialDims:
+      sdims = [Convolution.Dimension('HW', \
+          'Spatially packed HW. size#T#=H*W/strideW(#S0#). strideA#T#=strideW(#S0#).', \
+          DimAB.OnlyA)]
+    else:
+      sdims = []
+      schars = ['W','H','D']
+      for si in range(0,self.formatNumSpatialDims):
+        sc=schars[si]
+        sdims.append(Convolution.Dimension(sc,  \
+            'Spatial %s. size#T#=%s/stride%s(#S%d#) strideA#T#=stride%s(#S0#).'%(sc,sc,sc,si,sc), \
+            DimAB.OnlyA))
+    cdim = Convolution.Dimension('C',   'Cin.  size#T#=Cin.  stride#T#=1', DimAB.BothAB)
+
+    if convolutionType in ("ConvolutionForward", "ConvolutionBackwardData"):
+      # Set index assignment for output space
+      # Memory order for output space is NumCindices...0 with 0 the fastest-moving dim.
+      if self.tensorDFormat in ('NCHW','NCDHW'):
+        kidx = len(sdims)
+        nidx = kidx+1
+        cidx = nidx+1
+      elif self.tensorDFormat in ("CNHW", "CNDHW"):
+        nidx = len(sdims)
+        kidx = nidx+1
+        cidx = kidx+1
+
+      # Create summation dimensions for non-unit filters:
+      sumIdx = cidx
+      filterDims = []
+      for (fi,filterValue) in enumerate(self.filter[::-1]):
+        if filterValue != 1:
+          sumIdx = sumIdx+1
+          filterChar = chr(ord('Z')+self.formatNumSpatialDims-3-fi)
+          filterValueStr = "TBD" if filterValue==-1 else str(filterValue)
+          prevChar = ['1', 'W', 'W*H']
+          if self.dilation[fi] != -1:
+            dilationStr = str(self.dilation[fi])
+          else:
+            dilationStr = "TBD"
+          filterMsg = "Filter%s. size#T#=Filter%s(%s). strideA#T#=Dilation%s(%s)*%s." \
+              % (filterChar, filterChar, filterValueStr, filterChar, dilationStr, \
+                 prevChar[self.formatNumSpatialDims-fi-1])
+          filterDim = Convolution.Dimension(filterChar, filterMsg, DimAB.BothAB, strideA=self.dilation[fi])
+          filterDims.append( (sumIdx, Fbs.Sum, filterDim) )
+
+      spatialDims = []
+      for si,sdim in enumerate(sdims):
+        spatialDims.append((si, Fbs.Free, sdim))
+      self.numSpatialDims = len(spatialDims) # dims actually used in the tensor
+
+      if self.tensorAFormat in ("NCHW", "NCDHW"):
+        self.registerA( [(nidx,Fbs.Batch,ndim), (cidx,Fbs.Sum,cdim)] + spatialDims + filterDims )
+      elif self.tensorAFormat in ("NHWC", "NDHWC"):
+        self.registerA( [(nidx,Fbs.Batch,ndim)] + spatialDims + filterDims + [(cidx,Fbs.Sum,cdim)] )
+      elif self.tensorAFormat in ("CNHW", "CNDHW"):
+        self.registerA( [(cidx,Fbs.Sum,cdim), (nidx,Fbs.Batch,ndim)] + spatialDims + filterDims )
+      else:
+        raise RuntimeError ("unknown activation format '%s'"%self.tensorAFormat)
+
+      problemTypeOut["NumIndicesC"] = 2+len(spatialDims)
+
+      transposeCK =  convolutionType=="ConvolutionBackwardData"
+      ndim.strideB = 0
+      if self.tensorBFormat == "KCYX" and not transposeCK or\
```

Comment:
Could this translation be separated out into a function and unit tested?

Reply:
good idea